### ISSUE TITLE
ver0.9.9      espnet2/train/trainer.py:616  indenting error

### ISSUE THREAD

**Issue by zhangzhengyi-sogou** [espnet/espnet#3246]
An indenting error in 0.9.9 version  :
espnet2/train/trainer.py:616
The statement "optimizer.zero_grad()" should not be in the scope of "else" but in the scope of "if iiter % accum_grad == 0:"

**Comment by kamo-naoyuki** [comment]
You mean the follwoing line? 

https://github.com/espnet/espnet/blob/bed293ccd02d1fd462126911ad9975fcf973535b/espnet2/train/trainer.py#L616

It is in the scope of `"if iiter % accum_grad == 0:"`. See 

https://github.com/espnet/espnet/blob/bed293ccd02d1fd462126911ad9975fcf973535b/espnet2/train/trainer.py#L552

**Comment by zhangzhengyi-sogou** [comment]
I mean the following change:
else:


                 all_steps_are_invalid = False


                    with reporter.measure_time("optim_step_time"):


                        for iopt, (optimizer, scheduler) in enumerate(


                            zip(optimizers, schedulers)


                        ):


                            if optim_idx is not None and iopt != optim_idx:


                                continue


                            if scaler is not None:


                                # scaler.step() first unscales the gradients of


                                # the optimizer's assigned params.


                                scaler.step(optimizer)


                                # Updates the scale for next iteration.


                                scaler.update()


                            else:


                                optimizer.step()


                            if isinstance(scheduler, AbsBatchStepScheduler):


                                scheduler.step()


                            optimizer.zero_grad()
-----------------------------------------------------------------------------------

else:


                    all_steps_are_invalid = False


                    with reporter.measure_time("optim_step_time"):


                        for iopt, (optimizer, scheduler) in enumerate(


                            zip(optimizers, schedulers)


                        ):


                            if optim_idx is not None and iopt != optim_idx:


                                continue


                            if scaler is not None:


                                # scaler.step() first unscales the gradients of


                                # the optimizer's assigned params.


                                scaler.step(optimizer)


                                # Updates the scale for next iteration.


                                scaler.update()


                            else:


                                optimizer.step()


                            if isinstance(scheduler, AbsBatchStepScheduler):


                                scheduler.step()

optimizer.zero_grad()





发件人: Naoyuki Kamo ***@***.***
发送时间: 2021年5月28日 17:59
收件人: espnet/espnet
抄送: zhangzhengyi; Author
主题: Re: [espnet/espnet] ver0.9.9 espnet2/train/trainer.py:616 indenting error (#3246)


You mean the follwoing line?

https://github.com/espnet/espnet/blob/bed293ccd02d1fd462126911ad9975fcf973535b/espnet2/train/trainer.py#L616

It is in the scope of "if iiter % accum_grad == 0:". See

https://github.com/espnet/espnet/blob/bed293ccd02d1fd462126911ad9975fcf973535b/espnet2/train/trainer.py#L552

**Comment by zhangzhengyi-sogou** [comment]
![1622196798(1)](https://user-images.githubusercontent.com/77726109/119969051-a9f64280-bfe0-11eb-8dcb-cf4d9c87c813.png)

**Comment by kamo-naoyuki** [comment]
It's not in current version. Maybe you did it yourself it in the your espnet.

**Comment by Jarvan-Wang** [comment]
> 
> 
> It's not in current version. Maybe you did it yourself it in the your espnet.

No, because zhangzhengyi use old code
this bug exists before e51e7ec3fde8a21e1b65db5b6dd6fd68b99e1183
after 6bbe09501ef40d16dedc757ffdfebf3937f81761, this bug had been fixed since that.

**Comment by zhangzhengyi-sogou** [comment]
Yes,you got it.
